diffkemp/simpll/ValuePatternComparator.cpp

```diff
@@ -0,0 +1,80 @@
+//===---- ValuePatternComparator.cpp - Code pattern instruction matcher ---===//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Petr Silling, psilling@redhat.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the implementation of the value pattern matcher. The
+/// value pattern matcher is a comparator extension of the LLVM
+/// FunctionComparator tailored to value pattern comparison.
+///
+//===----------------------------------------------------------------------===//
+
+#include "ValuePatternComparator.h"
+#include "Config.h"
+
+/// Compare the given module value with the pattern value.
+int ValuePatternComparator::compare() {
+    auto PatternValue =
+            IsLeftSide ? ParentPattern->ValueL : ParentPattern->ValueR;
+
+    // Compare load from a global variable by name.
+    auto Load = dyn_cast<LoadInst>(ComparedValue);
+    if (Load && isa<GlobalVariable>(PatternValue)) {
+        return !namesMatch(Load->getOperand(0)->getName(),
+                           PatternValue->getName(),
+                           IsLeftSide);
+    }
+
+    // Compare all other values using the default implementation.
+    return cmpValues(ComparedValue, PatternValue);
+}
+
+/// Compare a module value with a pattern value without using serial numbers.
+/// Note: Parts of this function have been adapted from FunctionComparator.
+/// Therefore, LLVM licensing also applies here. See the LICENSE information
+/// in the appropriate llvm-lib subdirectory for more details.
+int ValuePatternComparator::cmpValues(const Value *ModVal,
+                                      const Value *PatVal) const {
+    // Catch self-reference case. Right side is the pattern side.
+    if (ModVal == FnL) {
+        if (PatVal == FnR)
+            return 0;
+        return -1;
+    }
+    if (PatVal == FnR) {
+        if (ModVal == FnL)
+            return 0;
+        return 1;
+    }
+
+    const Constant *ModConst = dyn_cast<Constant>(ModVal);
+    const Constant *PatConst = dyn_cast<Constant>(PatVal);
+    if (ModConst && PatConst) {
+        if (ModVal == PatVal)
+            return 0;
+        return cmpConstants(ModConst, PatConst);
```

Comment:
`cmpConstants` matches global values using `GlobalNumberState`, and unless they are compared earlier in `compare` when comparing loads, they get compared here this way. Is that really the desired behavior here? (I.e. is there a reason why treat load operations differently by comparing by name?)

Reply:
In value patterns, we match loads only by name, not by any other details (e.g., pointer type). I think this is necessary since value patterns should be defined by a single return instruction. More sophisticated mathing should use the generalist option, i.e., instruction patterns.

Reply:
I see, that should work fine then.